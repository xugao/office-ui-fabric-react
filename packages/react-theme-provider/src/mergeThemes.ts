import { PartialTheme, Theme } from './types';
import { merge } from '@uifabric/utilities';

/**
 * Merges all themes on top of a blank initial theme and ensures the theme is fully qualified.
 */
export const mergeThemes = (fullTheme: Theme, ...themes: (undefined | PartialTheme | Theme)[]): Theme => {
  const partialTheme = merge<PartialTheme | Theme>({}, fullTheme, ...themes);

  // Correctly merge stylesheets array
  // partialTheme.stylesheets = [];
  // themes.forEach(theme => theme && theme.stylesheets && partialTheme.stylesheets?.push(...theme.stylesheets));

  return partialTheme as Theme;
};
