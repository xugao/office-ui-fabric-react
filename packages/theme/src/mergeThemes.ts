import { merge } from '@uifabric/utilities';
import { IFontStyles, PartialTheme, Theme } from './types/index';
import { makeSemanticColors } from './utilities/makeSemanticColors';

export function mergeThemes(theme: Theme, partialTheme: PartialTheme = {}): Theme {
  const mergedTheme = merge<Theme | PartialTheme>({}, theme, partialTheme) as Theme;

  mergedTheme.semanticColors = makeSemanticColors(
    mergedTheme.palette,
    mergedTheme.effects,
    partialTheme?.semanticColors,
    !!partialTheme?.isInverted,
    false,
  );

  if (partialTheme.defaultFontStyle) {
    for (const fontStyle of Object.keys(mergedTheme.fonts) as (keyof IFontStyles)[]) {
      mergedTheme.fonts[fontStyle] = merge(mergedTheme.fonts[fontStyle], partialTheme.defaultFontStyle);
    }
  }

  if (partialTheme.stylesheets) {
    mergedTheme.stylesheets = (theme.stylesheets || []).concat(partialTheme.stylesheets);
  }

  return mergedTheme;
}
