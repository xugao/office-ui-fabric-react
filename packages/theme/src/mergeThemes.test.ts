import { mergeThemes } from './mergeThemes';

describe('mergeThemes', () => {
  it('can merge 2 themes', () => {
    expect(
      mergeThemes(
        {
          tokens: {
            body: {
              fill: 'blue',
              text: 'white',
            },
          },
        },
        { tokens: { body: { fill: 'red' } } },
      ),
    ).toMatchSnapshot();
  });
  it('can merge stylesheets', () => {
    expect(mergeThemes({ stylesheets: ['a', 'b'] }, { stylesheets: ['c', 'd'] })).toEqual({
      stylesheets: ['a', 'b', 'c', 'd'],
    });
  });
});
