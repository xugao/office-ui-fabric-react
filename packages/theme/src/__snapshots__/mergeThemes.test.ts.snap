// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeThemes can merge 2 themes 1`] = `
Object {
  "stylesheets": Array [],
  "tokens": Object {
    "body": Object {
      "fill": "red",
      "text": "white",
    },
  },
}
`;
